File: mesh_generator.py
"""
MeshGenerator

Description:
    A Python class for generating and visualizing 2D/3D meshes. This class provides
    functionalities to set up mesh parameters, generate mesh elements based on these
    parameters, and visualize the generated mesh using Matplotlib.

Author:
    JessyJP (2022) - Original MATLAB version
    Conversion to Python - [Your Name or Identifier] (2024)

License:
    GPLv3 - This code is released under the GNU General Public License v3.0.
    See LICENSE.md for more details.
"""

import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d.art3d import Poly3DCollection
from cube_voxel import CubeVoxel

def handle_single_or_list(method):
    def wrapper(elements, *args, **kwargs):
        if isinstance(elements, list):
            return [method(element, *args, **kwargs) for element in elements]
        else:
            return method(elements, *args, **kwargs)
    return wrapper


class MeshGenerator:
    def __init__(self):
        """
        Constructor for the MeshGenerator class.
        Initializes mesh parameters, the cube template for mesh elements, plotting settings,
        and other internal properties.
        """
        # Private properties (indicated by underscore prefix)
        self._cubeTemplate = None  # Cube prototype
        self._elementAlignmentTypes = None  # Element alignment types
        self._elements = []  # Store the elements
        self._xyz = None  # The dimension and setup of the mesh

        # Public properties
        self.cube = None  # Working cube modified
        self.dimRemoveID = 0  # 0=>OFF, 1=>x, 2=>y, 3=>z
        self.Pplot = {'View': 3, 'gridON': True, 'scaleFactor': 1, 'axisEqual': True,
                      'limX': [0, 1], 'limY': [0, 1], 'limZ': [0, 1]} # Plot settings
        self.fig = None  # Figure for plotting

        # Initialize internal and public properties
        self._initialize_properties()

    def _initialize_properties(self):
        """
        Initializes both private and public properties of the class.
        """
        # XYZ properties initialization
        self._xyz = {'x': {'Label': 'x', 'ID': 1, 'Lim': [0, 1], 'N': 1, 'gapP': 0.1, 'ElementL': None, 'gap': None, 'Coords': None, 'mesh': None},
                    'y': {'Label': 'y', 'ID': 2, 'Lim': [0, 1], 'N': 1, 'gapP': 0.1, 'ElementL': None, 'gap': None, 'Coords': None, 'mesh': None},
                    'z': {'Label': 'z', 'ID': 3, 'Lim': [0, 1], 'N': 1, 'gapP': 0.1, 'ElementL': None, 'gap': None, 'Coords': None, 'mesh': None}}
        # % Constant IDs
        # % Lim:      Defautl Limits
        # % N:        Default Number of Elements
        # % gapP:     Element Gap factor between elements
        # % ElementL: Element side length
        # % gap:      Element side length gap
        # % Coords:   Mesh grid coordinates
        # % mesh:     

        for key, axis in self._xyz.items():
            axis['ElementL'] = self._calc_el_spacing(axis) * (1 - axis['gapP'])
            axis['gap'] = self._calc_el_spacing(axis) * axis['gapP']
            axis['Coords'] = []
            axis['mesh'] = None

        # Initialize cube template using CubeVoxel
        self._cubeTemplate = CubeVoxel(id=0)

        # Initialize element alignment types
        self._elementAlignmentTypes = ["Symmetric", "Center"]

        # Copy the cube template to the public cube property
        self.cube = self._cubeTemplate.copy()

    def _calc_el_spacing(self, axis):
        """
        Calculate the spacing between elements along a given axis.

        Parameters:
        - axis: A dictionary containing the mesh properties for a specific axis.

        Returns:
        - The calculated element spacing as a float.
        """
        return (axis['Lim'][1] - axis['Lim'][0]) / axis['N']

    def setup_mesh(self, xyz_=None):
        """
        Sets up or updates the mesh based on provided or existing xyz parameters.
        
        Parameters:
        - xyz_: Optional dictionary containing updated mesh parameters for x, y, and/or z axes.
        """
        # Update the mesh parameters if a new xyz_ dictionary is provided
        if xyz_ is not None:
            self._xyz = xyz_.copy()
            print('Table updated!')
        else:
            print('Table not updated!')

        # Update cube template colors and face alpha from the cube
        self._cubeTemplate.colourF = self.cube.colourF
        self._cubeTemplate.colourE = self.cube.colourE
        self._cubeTemplate.faceAlpha = self.cube.faceAlpha
        # Apply the updated template to the cube
        self.cube = self._cubeTemplate.copy()

        # Update viewing limits based on xyz settings and scaleFactor % Default viewing Limitts 
        for axis_key, axis_props in self._xyz.items():
            self.Pplot[f'lim{axis_key.upper()}'] = np.array(axis_props['Lim']) * self.Pplot['scaleFactor']

        # Generate coordinates for each axis
        for axis_key, axis_props in self._xyz.items():
            self._xyz[axis_key] = self.setup_coordinate(self._xyz[axis_key])
            # axis_props['Coords'] = np.linspace(axis_props['Lim'][0], axis_props['Lim'][1], axis_props['N'] + 1)[:-1]

        # Handle dimension removal and adjust cube nodes if necessary
        if self.dimRemoveID > 0:
            # Adjust the cube's nodes based on the removed dimension
            for axis_key, axis_props in self._xyz.items():
                if axis_props['ID'] == self.dimRemoveID:
                    # This effectively removes the dimension by setting all coordinates in that dimension to 0
                    self.cube['nodesR'][:, self.dimRemoveID - 1] = 0

            # Adjust the FaceConnectNodes based on the removed dimension
            if self.dimRemoveID == 1:
                self.cube['FaceConnectNodes'] = np.array([[1, 2, 4, 3]]) - 1
            elif self.dimRemoveID == 2:
                self.cube['FaceConnectNodes'] = np.array([[1, 2, 3, 4]]) - 1
            elif self.dimRemoveID == 3:
                self.cube['FaceConnectNodes'] = np.array([[1, 2, 4, 3]]) - 1

        # Scale the cube based on the element spacing and gap percentage
        scale_factors = np.array([self._calc_el_spacing(self._xyz['x']) * (1 - self._xyz['x']['gapP']),
                                  self._calc_el_spacing(self._xyz['y']) * (1 - self._xyz['y']['gapP']),
                                  self._calc_el_spacing(self._xyz['z']) * (1 - self._xyz['z']['gapP'])])
        self.cube.scale( scale_factors)

        # Prepare elements for plotting
        X, Y, Z = np.meshgrid(self._xyz['x']['Coords'], self._xyz['y']['Coords'], self._xyz['z']['Coords'], indexing='ij')
        # Assign the generated mesh grid back to the _xyz attribute
        self._xyz['x']['mesh'] = X
        self._xyz['y']['mesh'] = Y
        self._xyz['z']['mesh'] = Z
        for i, (x, y, z) in enumerate(zip(X.flatten(), Y.flatten(), Z.flatten())):
            el = self.cube.copy()
            el.id = i
            # Translate the element based on the generated coordinates
            el.translate(np.array([x, y, z]));
            el.faceConnectNodes_Fast = el.faceConnectNodes + (i)* (8*(self.dimRemoveID==0)+4*(self.dimRemoveID!=0))
            self._elements.append(el)

    def setup_coordinate(self, axis):
        """
        Set up the coordinate array for a given axis based on its mesh properties.

        Parameters:
        - axis: A dictionary containing the mesh properties for a specific axis.

        Returns:
        - The axis dictionary updated with calculated element length, gap, and coordinates.
        """
        # Calculate element spacing (length) and gap based on the axis properties
        element_spacing = self._calc_el_spacing(axis) * (1 - axis['gapP'])
        gap = self._calc_el_spacing(axis) * axis['gapP']

        # Update the axis dictionary with calculated values
        axis['ElementL'] = element_spacing
        axis['gap'] = gap

        # Adjust the limits if the dimension corresponds to the one being removed
        if self.dimRemoveID == axis['ID']:
            axis['Lim'] = [0, 0]  # This effectively removes the dimension

        # Generate the coordinates for the axis
        coords = np.linspace(axis['Lim'][0], axis['Lim'][1], axis['N'] + 1)
        coords = coords[:-1]  # Exclude the last point to match MATLAB behavior

        # Update the axis dictionary with the generated coordinates
        axis['Coords'] = coords

        return axis


    def _adjust_cube_for_removed_dimension(self):
        """
        Adjusts the cube's nodes and face connections if a dimension is removed.
        """
        if self.dimRemoveID == 1:
            self.cube['FaceConnectNodes'] = np.array([[1, 2, 4, 3]]) - 1
        elif self.dimRemoveID == 2:
            self.cube['FaceConnectNodes'] = np.array([[1, 2, 3, 4]]) - 1
        elif self.dimRemoveID == 3:
            self.cube['FaceConnectNodes'] = np.array([[1, 2, 4, 3]]) - 1
        # Note: Additional logic may be needed to adjust nodesR based on the removed dimension.

    def _generate_mesh_elements(self):
        """
        Generates mesh elements based on the current settings of the mesh.
        """
        elements = []
        X, Y, Z = np.meshgrid(self._xyz['x']['Coords'], self._xyz['y']['Coords'], self._xyz['z']['Coords'], indexing='ij')
        for i, (x, y, z) in enumerate(zip(X.flatten(), Y.flatten(), Z.flatten())):
            el = self.cube.copy()
            el['id'] = i
            # Translate the element based on the generated coordinates
            el['nodesR'] += np.array([x, y, z])
            elements.append(el)
        return elements

    def plot_elements(self, elements):
        """
        Visualize the generated mesh elements using Matplotlib in a 3D plot.

        Parameters:
        - elements: Optional list of elements to plot. If None, plots all generated elements.
        """

        # Initialize the figure and axes if they don't exist or if the figure was closed
        if not hasattr(self, 'fig') or self.fig is None or not plt.fignum_exists(self.fig.number):
            self.fig = plt.figure()
            self.ax = self.fig.add_subplot(111, projection='3d')
            # Set the window style if needed; note that 'docked' style might not directly translate to matplotlib
        else:
            self.ax.clear()

        # Set plot limits and labels based on MeshGenerator properties
        self.ax.set_xlim(self.Pplot['limX'])
        self.ax.set_ylim(self.Pplot['limY'])
        self.ax.set_zlim(self.Pplot['limZ'])
        
        self.ax.set_xlabel('X')
        self.ax.set_ylabel('Y')
        self.ax.set_zlabel('Z')

        # Enable or disable the grid based on MeshGenerator properties
        self.ax.grid(self.Pplot.get('gridON', True))

        # Ensure axis aspect ratio is equal if specified
        if self.Pplot.get('axisEqual', True):
            self._set_equal_aspect()

        if not isinstance(elements, list):
            elements = [elements]
        # Plot each element
        for el in elements:
            # Assuming 'FaceConnectNodes' are indices starting from 0 in Python
            verts = [el.nodesR[face] for face in el.faceConnectNodes]
            face_collection = Poly3DCollection(verts, facecolors=el.colourF, edgecolors=el.colourE, alpha=el.faceAlpha)
            self.ax.add_collection3d(face_collection)

        plt.draw()
        plt.show()

    def plot_mesh_fast(self, elements):
        """
        Visualizes the mesh elements efficiently using a single Poly3DCollection.
        """
        if not hasattr(self, 'fig') or not plt.fignum_exists(self.fig.number):
            self.fig = plt.figure()
            self.ax = self.fig.add_subplot(111, projection='3d')
        else:
            self.ax.clear()

        # Prepare a list for collecting vertices and faces across all elements
        all_faces = []
        all_vertices = np.vstack([el['nodesR'] for el in elements])

        # Offset for vertices indexing in faces due to concatenation
        offset = 0
        for el in elements:
            # Adjust face indices based on current offset
            faces = el['FaceConnectNodes'] + offset
            all_faces.extend(faces)
            offset += len(el['nodesR'])

        # Create a Poly3DCollection from all faces and vertices
        face_collection = Poly3DCollection(all_vertices[all_faces], 
                                           facecolors=self.cube['colourF'], 
                                           edgecolors=self.cube['colourE'], 
                                           alpha=self.cube['FaceAlpha'])

        self.ax.add_collection3d(face_collection)

        # Set plot limits and labels based on MeshGenerator properties
        self.ax.set_xlim(self.Pplot['limX'])
        self.ax.set_ylim(self.Pplot['limY'])
        self.ax.set_zlim(self.Pplot['limZ'])
        
        self.ax.set_xlabel('X')
        self.ax.set_ylabel('Y')
        self.ax.set_zlabel('Z')

        # Enable or disable the grid based on MeshGenerator properties
        self.ax.grid(self.Pplot.get('gridON', True))

        # Ensure axis aspect ratio is equal if specified
        if self.Pplot.get('axisEqual', True):
            self._set_equal_aspect()

        plt.show()

    def _set_equal_aspect(self):
        """
        A workaround to set equal aspect ratio for the 3D plot, as Matplotlib does not support this directly.
        """
        extents = np.array([getattr(self.ax, f'get_{axis}lim')() for axis in 'xyz'])
        sz = extents[:,1] - extents[:,0]
        centers = np.mean(extents, axis=1)
        maxsize = max(abs(sz))
        r = maxsize/2
        for ctr, val in zip(centers, sz):
            self.ax.auto_scale_xyz(*[ctr - r, ctr + r], had_data=True)

    def update_colors_and_plot_properties(self, colorF, colorE, face_alpha, plot_scale_factor):
        """
        Updates the colors and plot properties of the mesh.

        Parameters:
        - colorF: Face color of the cube.
        - colorE: Edge color of the cube.
        - face_alpha: Transparency of the cube faces.
        - plot_scale_factor: Scale factor for the plot.
        """
        self.cube.colourF = colorF
        self.cube.colourE = colorE
        self.cube.faceAlpha = face_alpha
        self.Pplot['scaleFactor'] = plot_scale_factor

    def update_xyz_settings(self, xyz_settings):
        """
        Updates the xyz settings of the mesh based on the provided dictionary.

        Parameters:
        - xyz_settings: A dictionary containing the settings for x, y, and z axes.
        """
        for axis, settings in xyz_settings.items():
            if axis in self._xyz:
                for setting, value in settings.items():
                    if setting == 'Lim' and isinstance(value, list) and len(value) == 2:
                        self._xyz[axis]['Lim'] = value
                    elif setting in self._xyz[axis]:
                        self._xyz[axis][setting] = value
                # Recalculate element spacing, gap, and coordinates based on updated settings
                self._xyz[axis]['ElementL'] = self._calc_el_spacing(self._xyz[axis]) * (1 - self._xyz[axis]['gapP'])
                self._xyz[axis]['gap'] = self._calc_el_spacing(self._xyz[axis]) * self._xyz[axis]['gapP']
                self._xyz[axis]['Coords'] = self.setup_coordinate(self._xyz[axis])['Coords']
            else:
                print(f"Axis '{axis}' not found in mesh settings.")

    ## Modification methods
            
    @staticmethod
    @handle_single_or_list
    def translate(element, translation_vector):
        element.translate(translation_vector)
        return element

    @staticmethod
    @handle_single_or_list
    def scale(element, scale_factors):
        element.scale(scale_factors)
        return element

    @staticmethod
    @handle_single_or_list
    def rotate(element, angle, axis='z', center_of_rotation=None):
        element.rotate(angle, axis, center_of_rotation)
        return element

    @staticmethod
    @handle_single_or_list
    def transform_elements(element, transformation_function, S):
        element.transform_element(transformation_function, S)
        return element

    @staticmethod
    @handle_single_or_list
    def transform_nodes(element, transformation_function):
        element.transform_node(transformation_function)
        return element

    # Example usage of operator overloads
    def __mul__(self, matrix):
        for element in self.elements:
            element * matrix  # Assuming __mul__ is defined in CubeVoxel for matrix multiplication

    def __add__(self, vector):
        for element in self.elements:
            element + vector  # Assuming __add__ is defined in CubeVoxel for translation

    def __sub__(self, vector):
        for element in self.elements:
            element - vector  # Assuming __sub__ is defined in CubeVoxel for subtraction

    @property
    def xyz(self):
        """
        Property getter for the mesh dimensions.
        """
        return self._xyz

    def get_elements(self):
        return self._elements.copy()

    def get_matrix(self, elements, property):
        """
        Converts a specified property from a list of elements into a concatenated matrix.
        
        Parameters:
        - elements: List of elements to process.
        - property: The property of the elements to convert into a matrix.
        
        Returns:
        - A numpy array containing the concatenated values of the specified property from all elements.
        """
        return np.concatenate([el[property] for el in elements])

File: MeshGenerator.m
%
% Class-name:   MeshGenerator 
% Description: This script generates 2D/3D mesh and visualizes it.
% 
% Author: JessyJP (2022) % License: GPLv3 @ LICENCE.md
%

%% Mesh generator
classdef MeshGenerator < handle
    % MeshGenerator Class Implementation
    % This script generates 2D/3D mesh and visualizes it using CubeVoxel instances.
    
    % Internal properties
    properties (SetAccess=private)
        cubeTemplate;% This is a cube prototype
        elementAlignemntTypes;
        elements;% Store the elements
        xyz;% The dimension and setup of the mesh
    end

    % Public properties
    properties (Access=public)
        cube % Working cube modified
        dimRemoveID % 0=>OFF, 1=>x, 2=>y, 3=>z
        Pplot % Plot settings
        fig % Figure for plotting
    end
    
    methods
        % Constructor
        function [mesh] = MeshGenerator()
            % Settings
            mesh.dimRemoveID = 0;% 0=>OFF 1=>x 2=>y 3=>z
            % Labels
            xyz(1).Label='x';
            xyz(2).Label='y';
            xyz(3).Label='z';
            for i = 1:numel(xyz)
                % Constant IDs
                xyz(i).ID=i;
                % Mesh properties: Default Limits
                xyz(i).Lim = [0 1]; 
                % Mesh properties: Default Number of Elements
                xyz(i).N = 1;
                % Element Gap factor between elements
                xyz(i).gapP= 0.1;% Input 
                % Element side length
                xyz(i).ElementL = mesh.calcElSpacing(xyz(i))*(1-xyz(i).gapP); 
                % Element side length gap
                xyz(i).gap =  mesh.calcElSpacing(xyz(i))*(xyz(i).gapP);
                % Mesh grid coordinates
                xyz(i).Coords = {};
                % Mesh grid coordinates
                xyz(i).mesh =[];
            end

            % Element Alignment
            mesh.elementAlignemntTypes = ["Symetric";"Center"];
            % Default Plot view direction
            mesh.Pplot = struct( ...
                'View', 3, 'gridON', true, 'scaleFactor', 1, 'axisEqual', true, ...
                'limX', [0, 1], 'limY', [0, 1], 'limZ', [0, 1]);


            % Assign Mesh
            mesh.xyz = struct2table(xyz,'RowNames',{xyz.Label});


            % Initialize cube template with default CubeVoxel
            mesh.cubeTemplate = CubeVoxel(0, [], [], [], [], 1.0);
            mesh.cube = mesh.cubeTemplate; % Copy the cube template to the public cube property            
        end
        
        % Setup mesh
        function setupMesh(mesh,xyz_)% mesh obj
            % Update the mesh parameters 
            if nargin == 2 && istable(xyz_)
                mesh.xyz = xyz_;
%                 disp(mesh.xyz);
                disp('Table updated!');
            else
                disp('Table not updated!');
            end
            
            % Setup Properties and Update colours
            mesh.cubeTemplate.updateColourPropertiesFrom(mesh.cube);
            % Update the colours of the template cube
            mesh.cube = mesh.cubeTemplate;
            
            % Default viewing Limits 
            mesh.Pplot.limX = mesh.xyz{'x','Lim'}*mesh.Pplot.scaleFactor;
            mesh.Pplot.limY = mesh.xyz{'y','Lim'}*mesh.Pplot.scaleFactor;
            mesh.Pplot.limZ = mesh.xyz{'z','Lim'}*mesh.Pplot.scaleFactor;                       
            
            % Make coordinates
            mesh.xyz('x',:) = mesh.setupCoordinate(mesh.xyz('x',:));
            mesh.xyz('y',:) = mesh.setupCoordinate(mesh.xyz('y',:));
            mesh.xyz('z',:) = mesh.setupCoordinate(mesh.xyz('z',:));

            % Remove the dimension component 
            if mesh.dimRemoveID > 0 && mesh.dimRemoveID <= 3
                % Remove the cube nodes that correspond to the removed dimension   
                ind = not(mesh.cube.nodesR(:,mesh.dimRemoveID) > 0);
                mesh.cube.nodesR = mesh.cube.nodesR(ind,:);

                % Get the right square face
                % For removing For X
                if mesh.dimRemoveID == 1                    
                    mesh.cube.FaceConnectNodes = [1 2 4 3];
                end
                % For removing For Y
                if mesh.dimRemoveID == 2                    
                    mesh.cube.FaceConnectNodes = [1 2 3 4];
                end
                % For removing For Z
                if mesh.dimRemoveID == 3
                    mesh.cube.FaceConnectNodes = [1 2 4 3];
                end
            end
            
            % Scale the cube
            mesh.cube = mesh.scale(mesh.cube,...
                [mesh.calcElSpacing(mesh.xyz('x',:))  ...
                 mesh.calcElSpacing(mesh.xyz('y',:)) ...
                 mesh.calcElSpacing(mesh.xyz('z',:))].*(1-mesh.xyz.gapP'), ...
                 [0 , 0, 0]);

            % Cube Test preview
%             mesh.PplotElements(mesh.translate(mesh.cube,[-1 -1 -1]*0.5*0));
            
            [X,Y,Z] = meshgrid(mesh.xyz{"x","Coords"}{:},mesh.xyz{"y","Coords"}{:},mesh.xyz{"z","Coords"}{:});
            % Assign Mesh
            mesh.xyz{"x","mesh"} = {X};
            mesh.xyz{"y","mesh"} = {Y};
            mesh.xyz{"z","mesh"} = {Z};
            
            Elements_ = [];
            offsetNodeInd = 0;
            % Make mesh
            for i = 1:numel(X)
                % Make elements
                el = mesh.cube.deepCopy();
                el.id = i;
                el.offsetNodeInd  = offsetNodeInd ;                
                % Update the face node index                
                offsetNodeInd = offsetNodeInd  + size(el.nodesR,1);
                
                % Translate the 
                el = el.translate([X(i),Y(i),Z(i)]);
                % For fast plotting
                el.FaceConnectNodes_Fast = el.FaceConnectNodes + (i-1)*(8*(mesh.dimRemoveID==0)+4*(mesh.dimRemoveID~=0)); 
%                 El.colourF_Fast     = repmat(El.colourF,   size(El.FaceConnectNodes_Fast,1),1);
%                 El.FaceAlpha_Fast  = repmat(El.FaceAlpha,size(El.FaceConnectNodes_Fast,1),1);
                % Gather elements together
                Elements_ = [Elements_;el];
            end
            mesh.elements = Elements_;
            % Plot Preview
%             mesh.PplotElements(mesh.translate(Elements_,[0 0 0]*0.5*0));


        end
                
        % Setup Coordinate
        function [c] = setupCoordinate(mesh,c)%s - general coordinates, P - properties

            % Setup Mesh properties: Element length
            c.ElementL = mesh.calcElSpacing(c)*(1-c.gapP);
            % Setup Mesh properties: Element gap
            c.gap      = mesh.calcElSpacing(c)*(c.gapP);

            % Remove a dimension for the grid
            c.Lim = c.Lim*not(mesh.dimRemoveID==c.ID);

            % Make a grid origins i.e. the [0 0 0] vertex of the Element cube polygon
            Coords = linspace(c.Lim(1),c.Lim(2),(c.N+1));
            Coords(end) =[]; 
            c.Coords = {Coords};
        end
        
        function size_element = calcElSpacing(~,c)
        % function size_element = sizeElement(~,c)
            size_element = (c.Lim(2)-c.Lim(1))./c.N;
        end
            
        % Subfunction for plotting element
        function plotElements(mesh,elements, plotEdgeAlpha)
            % Get handles and do plotting
            try 
                if not(isvalid(mesh.fig))
                    error();
                end
            catch
                mesh.fig = figure();
                mesh.fig.WindowStyle = 'docked';
            end
            clf(mesh.fig);
            view(mesh.Pplot.View);
            ax = mesh.fig.Children(1);
                
            if nargin < 3
                plotEdgeAlpha = false;
            end

            % Get axes. Setup plot limits and labels
%             ax = mesh.ax;
            if mesh.Pplot.axisEqual
                axis(ax,'equal');
            end
            ax.XLim = mesh.Pplot.limX;
            ax.YLim = mesh.Pplot.limY;
            ax.ZLim = mesh.Pplot.limZ;
            ax.XLabel.String = 'x';
            ax.YLabel.String = 'y';
            ax.ZLabel.String = 'z';
            if mesh.Pplot.gridON
                grid(ax,'on')
            end
            mesh.Pplot.axP = [];
            
            N_elements = numel(elements);
            % Plot elements one by one
            for i = 1:N_elements
                E = elements(i);

                if (not(plotEdgeAlpha))
                    mesh.Pplot.axP(i) = patch(ax,'Faces', E.FaceConnectNodes, 'Vertices', E.nodesR, ...
                    'Facecolor', E.colourF,'FaceAlpha',E.faceAlpha,'EdgeColor',E.colourE); 
            %         drawnow
                else
                    % Plot faces
                    mesh.Pplot.axP(i) = patch(ax, 'Faces', E.FaceConnectNodes, 'Vertices', E.nodesR, ...
                          'Facecolor', E.colourF, 'FaceAlpha', E.faceAlpha, 'EdgeColor', 'none');
                    
                    % Plot edges with specified edge alpha
                    edgeAlpha = E.edgeAlpha; % Example edge alpha value
                    for j = 1:size(E.FaceConnectNodes, 1)
                        face = E.FaceConnectNodes(j, :);
                        for k = 1:length(face)
                            % Determine the next vertex in the face (or loop back to the start)
                            nextIdx = mod(k, length(face)) + 1;
                            % Extract the coordinates for the current and next vertex
                            v1 = E.nodesR(face(k), :);
                            v2 = E.nodesR(face(nextIdx), :);
                            % Plot the edge as a line with specified transparency
                            line(ax, [v1(1), v2(1)], [v1(2), v2(2)], [v1(3), v2(3)], ...
                                 'Color', [E.colourE, edgeAlpha], 'LineWidth', 1.5);
                        end
                    end                    
                end
                printProgressBar(i,N_elements,"barLength",20)
            end
            drawnow;
        end

        % Subfunction for plotting element
        function plotMeshFast(mesh,elements)
            % Get handles and do plotting
            try 
                if not(isvalid(mesh.fig))
                    error;
                end
            catch
                mesh.fig = figure();
                mesh.fig.WindowStyle = 'docked';
            end
            clf(mesh.fig);
            view(mesh.Pplot.View);
            ax = mesh.fig.Children(1);
            % Get axes and setup plot limits
%             ax = mesh.ax;
            if mesh.Pplot.axisEqual
                axis(ax,'equal');
            end
            ax.XLim = mesh.Pplot.limX;
            ax.YLim = mesh.Pplot.limY;
            ax.ZLim = mesh.Pplot.limZ;
            ax.XLabel.String = 'x';
            ax.YLabel.String = 'y';
            ax.ZLabel.String = 'z';
            if mesh.Pplot.gridON
                grid(ax,'on')
            end
            mesh.Pplot.axP = [];
            
            % Get properties ax matrix
            FaceConnectNodes_Fast = mesh.getMatrix(elements,"FaceConnectNodes_Fast");            
            nodesR = mesh.getMatrix(elements,"nodesR");
%             colour =  mesh.getMatrix(elements,"Colour_Fast");
%             faceAlpha = mesh.getMatrix(elements,"FaceAlpha_Fast");
            
            mesh.Pplot.axP = ...
            patch(ax,'Faces', FaceConnectNodes_Fast, 'Vertices', nodesR,...
            'FaceColor',mesh.cube.colourF,'FaceAlpha',mesh.cube.faceAlpha, 'EdgeColor', mesh.cube.colourE); 
            drawnow;
        end
        
        % Subfunction for translating element
        function [elements] = translate(~,elements,translationVector)
            for i = 1:numel(elements)
                elements(i) =  elements(i).translate(translationVector);
            end
        end

        % Subfunction for scaling element
        function [elements] = scale(~, elements, scaleFactors, centerPoint)
            % Applies scaling to each voxel in the mesh.
            % mesh: Array or cell array of CubeVoxel instances.
            % scaleFactors: A 3-element vector specifying scale factors along x, y, z axes.
            % centerPoint: Optional. A 3-element vector specifying the center point to scale around.
            %              If not provided, each voxel's center is used.
        
            if nargin < 4
                useVoxelCenter = true;
            else
                useVoxelCenter = false;
            end
        
            for i = 1:length(elements)
                if useVoxelCenter
                    elements(i) = elements(i).scale(scaleFactors);
                else
                    elements(i) = elements(i).scale(scaleFactors, centerPoint);
                end
            end
        end

        function [elements] = rotate(~, elements, angle, axis, pivotPoint)
            % Applies rotation to each voxel in the mesh.
            % mesh: Array or cell array of CubeVoxel instances.
            % angle: Rotation angle in degrees.
            % axis: 'x', 'y', or 'z' axis to rotate around.
            % pivotPoint: Optional. A 3-element vector specifying the pivot point.
            %             If not provided, each voxel's center is used.
        
            if nargin < 5
                useVoxelCenter = true;
            else
                useVoxelCenter = false;
            end
        
            for i = 1:length(elements)
                if useVoxelCenter
                    elements(i) = elements(i).rotate(angle, axis);
                else
                    elements(i) = elements(i).rotate(angle, axis, pivotPoint);
                end
            end
        end


        % Transformation functions
        function [elements] = transformElements(~,elements,transformationF)
        % function [elements] = transformElements(~,elements,transformationF)   
            for i = 1:numel(elements)
                % Translate vector        
                elements(i) = transformationF(elements(i));                
            end
        end
        
        function [elements] = transformNodesR(~,elements,transformationF)
        % function [elements] = transformNodesR(~,elements,transformationF)   
            for i = 1:numel(elements)
                % Transform Nodes
                elements(i).nodesR = transformationF(elements(i).nodesR);                
            end
        end
        
        function elementsCopy = copyElements(mesh)
            % Creates a deep copy of all elements in the mesh.
            % Returns an array of CubeVoxel instances that are deep copies of the original elements.
            
            % Initialize an empty array to hold the copies.
            % Depending on how elements are stored, you might need to use {} for a cell array.
            elementsCopy = repmat(CubeVoxel(0, [], [], [], [], 1.0), size(mesh.elements));
            
            % Iterate over each element and create a deep copy.
            for i = 1:length(mesh.elements)
                elementsCopy(i) = mesh.elements(i).deepCopy();
            end
        end
        
        % Get method for the max XYZ
        function xyz = get.xyz(mesh)
             xyz = mesh.xyz;
        end
        
        % Convert elements to a matrix of coordinates
        function M = getMatrix(~,elements,property)
        % function M = getMatrix(~,elements,property)
        % M = cell2mat({elements.(property)}');
            M = cell2mat({elements.(property)}');
        end
        
        function vertexMatrix = elementsToVertexMatrix(mesh)
            % Extract all nodes from each element and concatenate them into a single matrix
            allNodes = vertcat(mesh.elements.nodesR);
            
            % Assuming each vertex has a unique ID across the mesh, sort or organize
            % the vertices if necessary (this step may adjust based on how IDs are assigned and used)
            % For simplicity, this example assumes vertices are already in the correct order
            
            vertexMatrix = allNodes; % Each row is [x, y, z] for a vertex
        end

        function vertexMatrixToElements(mesh, vertexMatrix)
            % Starting index for vertices in the matrix
            startIndex = 1;
            
            for i = 1:length(mesh.elements)
                % Number of vertices for the current element
                numVertices = size(mesh.elements(i).nodesR, 1);
                
                % Extract the subset of the vertex matrix for the current element
                endIndex = startIndex + numVertices - 1;
                mesh.elements(i).nodesR = vertexMatrix(startIndex:endIndex, :);
                
                % Update the start index for the next element
                startIndex = endIndex + 1;
            end
        end


        function exportToFile(mesh, filename, format)
            % Validate the format
            validFormats = ["OBJ", "STL"];
            if ~ismember(upper(format), validFormats)
                error(['Unsupported format: ', format]);
            end
            
            % Ensure filename ends with the correct extension
            [path, name, ext] = fileparts(filename);
            if ~strcmpi(ext, ['.', lower(format)])
                filename = fullfile(path, join([name, '.', lower(format)],""));
            end
            
            % Open the file
            fileID = fopen(filename, 'w');
            if fileID == -1
                error(['Failed to open file: ', filename]);
            end
            
            % Write the file header based on the format
            switch upper(format)
                case "OBJ"
                    fprintf(fileID, '# OBJ file exported from MATLAB\n');
                case "STL"
                    fprintf(fileID, 'solid mesh\n');
            end
            
            % Write the mesh data
            for i = 1:numel(mesh.elements)
                switch upper(format)
                    case "OBJ"
                        fprintf(fileID, '%s', mesh.elements(i).toOBJString());
                    case "STL"
                        fprintf(fileID, '%s', mesh.elements(i).toSTLString());
                end
                printProgressBar(i,numel(mesh.elements),"barLength",20);
            end
            
            % Write the file footer based on the format
            if upper(format) == "STL"
                fprintf(fileID, 'endsolid mesh\n');
            end
            
            % Close the file
            fclose(fileID);
        end

    end
end


